lib/src/tree_cursor.c

```diff
@@ -264,25 +269,61 @@ TSFieldId ts_tree_cursor_current_field_id(const TSTreeCursor *_self) {
       }
     }
 
-    const TSFieldMapEntry *field_map, *field_map_end;
+    if (ts_subtree_extra(*entry->subtree)) break;
+
     ts_language_field_map(
       self->tree->language,
       parent_entry->subtree->ptr->production_id,
-      &field_map, &field_map_end
+      field_map, field_map_end
     );
-
-    while (field_map < field_map_end) {
-      if (
-        !ts_subtree_extra(*entry->subtree) &&
-        !field_map->inherited &&
-        field_map->child_index == entry->structural_child_index
-      ) return field_map->field_id;
-      field_map++;
+    for (const TSFieldMapEntry *i = *field_map; i < *field_map_end; i++) {
+      if (!i->inherited && i->child_index == entry->structural_child_index) {
+        *child_index = entry->structural_child_index;
+        return i->field_id;
+      }
     }
   }
   return 0;
 }
 
+TSFieldId ts_tree_cursor_current_field_id_ext(
+  const TSTreeCursor *self,
+  bool *field_has_additional
+) {
+  uint32_t child_index;
+  const TSFieldMapEntry *field_map, *field_map_end;
+  TSFieldId field_id = ts_tree_cursor__current_field_info(
+    self,
+    &field_map,
+    &field_map_end,
+    &child_index
+  );
+
+  // After finding the field, check if any other later children have
+  // the same field name.
+  if (field_id) {
+    for (const TSFieldMapEntry *i = field_map; i < field_map_end; i++) {
+      if (i->field_id == field_id && i->child_index > child_index) {
+        *field_has_additional = true;
+      }
+    }
+  }
+
+  return field_id;
+}
```

Comment:
This is kind of a weird new internal function. It returns the field associated with the current node, and it also indicates (via the `*field_has_additional` out-parameter) whether any other, later siblings of this node can possibly have the *same* field.

This allows for an important optimization in the pattern-matching process. Most of the time, field names are single-valued: they are only associated with one child node. For a pattern like this:

```clojure
(method_definition
  name: (identifier) @the-function-name)
```

when we find the `identifier` node with the `name` field, we can know for sure that the *only* way that this pattern can match is for it to capture the current node; there will not be any other `identifier` nodes with the `name` field. This allows us to update our `QueryState` for this pattern *in-place*, instead of splitting it as we would normally need to do when adding a capture, in order to perform a truly exhaustive search.